### ISSUE TITLE
Use TiDB instead of MySQL as default database

### ISSUE THREAD

**Issue by lastzero** [photoprism/photoprism#60]
[<img alt="Issuehunt badges" src="https://img.shields.io/badge/IssueHunt-%2460%20Rewarded-%237E24E3.svg" />](https://issuehunt.io/r/photoprism/photoprism/issues/60)

**As a potential user, I don't want to run a full blown MySQL server in order to simplify installation and reduce memory usage.**

We solved this by replacing MySQL with TiDB, a compatible database server implemented in pure Go. Most of the changes are done! Our [demo](https://demo.photoprism.org/) already runs on TiDB.

It starts automatically, if `tidb` is selected as database-driver. Initialization and configuration should be improved, still a bit experimental at the moment. Also we need test coverage and some code clean-up.

Acceptance Criteria:
- [x] TiDB MUST be initialized with `photoprism` as user, password and database without restart; using `FLUSH PRIVILEGES` should work (see [comment](https://github.com/photoprism/photoprism/issues/60#issuecomment-448834543))
- [x] TiDB root password MUST be `photoprism` (can be changed by user during setup later, not part of this ticket)
- [ ] Integration tests SHOULD run on TiDB as well (MySQL still used right now, see [docker-compose.yml](https://github.com/photoprism/photoprism/blob/develop/docker-compose.yml)); started working on `photoprism/tidb` Docker image for this
- [x] Code clean-up in `/internal/tidb` SHOULD be performed
- [x] Unit tests MUST be added
- [x] We should ask the TiDB developers if they have any concerns.

---

<details>
<summary>
<b>IssueHunt Summary</b>
</summary>

### Backers (Total: $60.00)

- [<img src='https://avatars3.githubusercontent.com/u/44827199?v=4' alt='issuehunt' width=24 height=24> issuehunt](https://issuehunt.io/u/issuehunt) ($60.00)

### Submitted pull Requests
- [#129 WIP: POC for using go-sqlmock for db interaction #84](https://issuehunt.io/r/photoprism/photoprism/pull/129)
---

### Tips

- Checkout the [Issuehunt explorer](https://issuehunt.io/r/photoprism/photoprism/) to discover more funded issues.
- Need some help from other developers? [Add your repositories](https://issuehunt.io/r/new) on IssueHunt to raise funds.
---
IssueHunt has been backed by the following sponsors. [Become a sponsor](https://issuehunt.io/membership/members)
</details>

**Comment by lastzero** [comment]
We are now definitively working on a [single binary for distribution](https://github.com/photoprism/photoprism/wiki/Distribution). Many users certainly don't want to install Docker (and keep it updated).

**Comment by robmarkcole** [comment]
@lastzero how can I help? I am a python dev but very keen on supporting this project

**Comment by lastzero** [comment]
@robmarkcole Thank you for reaching out! A good start is to join our developers mailing list: https://groups.google.com/a/photoprism.org/forum/#!forum/developers

This is a young project. We are in the process of organizing work and building a community. We have no issues that require Python at the moment (possibly later). Maybe you're interested in doing research (like testing different RAW to JPEG converters) or learning Go?

**Comment by lastzero** [comment]
From what I learned, it requires significant work to use sqlite3 concurrently. Before we start refactoring, let's see if we can run a MySQL- or PostgreSQL-compatible server within PhotoPrism:

- [TiDB](https://pingcap.com) - a distributed HTAP database compatible with the MySQL protocol
- [CockroachDB](https://github.com/cockroachdb/cockroach/) - ultra-resilient SQL for global business (PostgreSQL protocol)

It seems like a crazy idea, but why not?

**Comment by robmarkcole** [comment]
HI @lastzero I am interested in learning Go, and know some SQL. In my day job I work with satellite imagery, and several services I use have PostgreSQL as a back-end. If theres a task in this area that will get me using Go please let me know :)

**Comment by webrules** [comment]
I had a look at TiDB. It can be an (almost) drop-in replacement for MySQL. However, only Docker or manual installations on Linux are available. Guess it may not help on the vision of single binary distribution of PhotoPrism?

**Comment by lastzero** [comment]
@webrules It seems like you can build TiDB on Windows: https://github.com/pingcap/tidb/issues/4295

I would actually not build a separate binary but compile this into our own. Should be possible, as long as there are no external dependencies, like shared libraries that are only available on Linux.

**Comment by webrules** [comment]
A bit confused. Does that mean you want to replace MySQL in the docker image with TiDB or CockroachDB? Or it is more than that?

**Comment by lastzero** [comment]
We don't want to use Docker anymore. Essentially, PhotoPrism would be its own database server.

**Comment by lastzero** [comment]
[File Browser](https://filebrowser.github.io) seems to use SQLite3 - we should check their implementation.

**Comment by jed-frey** [comment]
I would suggest using an ORM so that the backend can be configurable.

http://xorm.io/

**Comment by lastzero** [comment]
Most of the changes are basically done! Our [demo](https://demo.photoprism.org/) already runs on TiDB.

- [ ] Would be amazing if TiDB could be initialized with a user and database without restart!
- [ ] Integration tests still run on MySQL. Need to figure out how to replace it with TiDB - that's why easy initialization is important, like we can do it with the MySQL image. Started working on a `photoprism/tidb` Docker image for this.
- [ ] Code clean-up in `/internal/tidb` and unit tests are required.
- [ ] We should ask the TiDB developers if they have any concerns.

**Comment by lastzero** [comment]
@c4pt0r Thank you! What can happen in the worst case? This software will mainly run on desktop computers and other small devices, so we can't use a cluster by default. We will keep metadata in XMP sidecar files, so a database crash is not the end of the world.

**Comment by jed-frey** [comment]
> This software will mainly run on desktop computers and other small devices.

Has that design decision been made? While I am an atypical user, I've been shooting digital photos since 2002. I have ~1.5TB of pictures all archived on my NAS. I'm slowly hacking through [my own solution in python](https://github.com/jed-frey/python_PhotoIndexer), but I can run locally but scale to every core in my house to scan EXIF and images.

**Comment by lastzero** [comment]
@jed-frey A NAS is more similar to a desktop computer than to typical cloud computing environments with lots of nodes far away from home. There is a tradeoff between scalability and complexity. Thank god, modern computers - even smartphones - are very powerful and can easily handle 1.5 TB, given enough storage capacity.

**Comment by c4pt0r** [comment]
> @c4pt0r Thank you! What can happen in the worst case? This software will mainly run on desktop computers and other small devices, so we can't use a cluster by default. We will keep metadata in XMP sidecar files, so a database crash is not the end of the world.

Cool, I got your point, I think it's not a big deal. TiDB uses goleveldb as the underlying storage for single node mode, and it's solid enough for desktop application. And I think it's pretty cool if you can put tidb and photoprism into one binary.

**Comment by lastzero** [comment]
@c4pt0r It would be amazing if we could initialize TiDB with a user, password and database without restart. Either we add this functionality to TiDB somehow (what is the best way?), or we ship our app with existing database files. Would they be binary compatible for all platforms (processors and operating systems)?

**Comment by c4pt0r** [comment]
> @c4pt0r It would be amazing if we could initialize TiDB with a user, password and database without restart. Either we add this functionality to TiDB somehow (what is the best way?), or we ship our app with existing database files. Would they be binary compatible for all platforms (processors and operating systems)?

I vote for shipping your app with existing database files (or create schema statements), it's a more flexible design.

**Comment by c4pt0r** [comment]
> @c4pt0r It would be amazing if we could initialize TiDB with a user, password and database without restart. Either we add this functionality to TiDB somehow (what is the best way?), or we ship our app with existing database files. Would they be binary compatible for all platforms (processors and operating systems)?

I think shipping with existing database file is not a good idea, I suggest to putting create table statements with your application, execute them when bootstrapping. It's a more flexible way.

**Comment by lastzero** [comment]
@c4pt0r Why did you change your mind? The problem with create statements is, that the client doesn't know if they've already been executed plus the server could be attacked as root without password for at least a short period of time. This is a quick hack to create a database:

https://github.com/photoprism/photoprism/blob/develop/internal/tidb/init.go

Changing the root password or other user credentials however only seems to take effect after server restart, that's why I removed it again. Bug or feature? Thank you very much for your advice 👍

**Comment by c4pt0r** [comment]
You need to run 'FLUSH PRIVILEGES;' to make it work, instead of restarting the server.

MYSQL> SET PASSEWORD='new_password';
MYSQL> FLUSH PRIVILEGES;

**Comment by lastzero** [comment]
Ah, great. Thought it is not implemented because of https://github.com/pingcap/tidb/issues/674. Top position when you google for "FLUSH PRIVILEGES tidb".

https://github.com/pingcap/docs/blob/master/sql/user-account-management.md also doesn't mention it. I'll send a PR.
